[OptimistLabyrinth] 🚀 3단계 - 로또(자동)

OptimistLabyrinth opened the pull request:
Controller, Model, View 로 분리해서
각각의 역할에 맞게 실행하도록 최대한 맞춰서 구현해봤습니다.

구현하는 내내 테스트 코드를 되도록 거의 동시에 작성하면서 작업하려고 노력했어요.
결과적으로 Model 쪽에 단위 테스트를 상당히 충실하게 작성할 수 있게 되었고
작업을 진행하는 중간 중간에 리팩터링이 필요한 경우에도
코드 변경하는게 굉장히 편하다고 느낄만큼 테스트 코드를 작성하는 것의 장점을 많이 느끼는 시간이었어요 🙂😄

시간 되실때 피드백 부탁드리겠습니다 ㅎㅎ

Commit:
docs: 3단계 로또(자동) 요구사항 문서로 정리

Commit:
feat: 로직을 정리하며 필요한 클래스를 생성한다

Commit:
feat: 구입 금액을 사용자로부터 입력 받아서 적절한 형태로 저장하기

MoneyToBuyAcceptor 클래스 구현
MoneyToBuy 클래스 구현

MoneyToBuy 에 대한 단위 테스트 작성

Commit:
feat: 구입 금액 입력값에서 로또 구입 금액으로 나눈 나머지는 제거한다

테스트 코드도 함께 작성

Commit:
feat: 구입 금액으로 구매 가능한 로또 장 수를 확인하는 메서드 추가

MoneyToBuy 클래스 numberOfAffordableLottoTickets 메서드 구현

Commit:
refactor: 로또 한 장 가격을 이름있는 상수로 만들기

Commit:
feat: 메인 로직 정리한 로직 수정하기

Commit:
feat: 구매 완료한 로또 장 수를 출력하는 로직 구현

NumberOfLottoTicketsPrinters 클래스

Commit:
feat: 1 이상 45 이하의 무작위 번호를 생성하는 로직 구현

LottoNumberGenerator 클래스

단위 테스트 작성

Commit:
feat: 로또 한 장을 만들어내는 로직 구현

LottoTicket 클래스 구현

Commit:
feat: 여러 장의 로또 티켓을 관리하기 위한 클래스 구현

LottoTicketsBucket 클래스 구현

Commit:
fix: LottoNumberGenerator 생성자에서 무작위 선택 리스트 제공하도록 변경

무작위 숫자 선택 시 중복을 제거하는 로직을 구현하기 위해서
LottoNumberGenerator 클래스에게 선택 리스트를 외부에서 주입해야 한다

Commit:
feat: 한 장의 로또에 있는 6 개의 숫자에 중복이 없도록 수정

LottoTicket 클래스 수정
LottoTicket 에 대한 단위 테스트 작성

Commit:
fix: 중첩 클래스(nested class) 는 반드시 static 표시하기

Commit:
refactor: 불필요하게 리스트 정렬하는 부분 개선

Commit:
chore: 멤버 변수 직접 접근도 가능하기는 하지만 컨벤션에 따라서 getter 사용하기

Commit:
feat: LottoTicket 클래스에서 랜덤에 대한 의존도를 낮추기 위해서 DI 사용하도록 수정

기존에 LottoTicket 클래스에서 직접적으로 생성해서 사용하던 의존성 객체를
생성자를 통해 주입받도록 수정해서 코드 결합도(coupling)를 낮춘다

LottoNumberGenerator 클래스에 대한 단위 테스트 코드도 함께 수정한ㄷ
LottoTicket 클래스에 대한 단위 테스트 코드도 함께 수정한다

Commit:
test: LottoTicketsBucket 클래스에 대한 단위 테스트 작성

작성하는 과정에서 LottoTicketsBucket 생성자 내부 구현에서 오류 발견 후 수정

Commit:
feat: Controller 에서의 사용 방식을 고려해서  LottoTicketsBucket 생성자 수정

외부에서 LottoTickets 객체를 만들어서 사용한 후에
나중에 저장 및 관리를 위해서 LottoTicketsBucket 객체에 넣어주게 된다

실제 사용 방식을 LottoTicketsBucket 클래스에 대한 단위 테스트에서 확인 가능

Commit:
feat: 구매한 로또를 한 장 씩 출력한다

LottoTicket 클래스에서 toString 메서드 오버라이드

Commit:
feat: 빈 줄 한 줄 출력한다

Commit:
feat: 사용자에게 당첨 번호를 입력하도록 메시지를 출력한다

Commit:
feat: UI 에서 사용자에게 당첨 번호를 입력 받는다

입력 받은 결과를 WinningNumbers 타입 객체로 래핑해서 반환한다

Commit:
feat: 개별 당첨 번호를 저장하고 관리하는 클래스 구현

WinningNumberEach 클래스 구현

WinningNumberEach 에 대한 단위 테스트 작성

Commit:
test: WinningNumbers 클래스에 대한 단위 테스트 작성

WinningNumbers 클래스에 대한 단위 테스트 코드를 작성하면서
WinningNumbers 내에서 부족한 로직을 채워나감

추가적으로 WinningNumberEach 클래스에도 필요한 부분을 추가함

Commit:
refactor: isNullOrEmpty 메서드 별도로 분리하기

constant/utils/StringUtils 클래스로 분리

Commit:
refactor: 이름있는 상수 constant/numbers 폴더에 몰아놓기

로또 개별 번호 최솟값: LOTTO_MINIMUM_NUMBER
로또 개별 번호 최댓값: LOTTO_MAXIMUM_NUMBER
로또 한 장에 들어있는 숫자 개수: COUNT_OF_NUMBERS_IN_SINGLE_LOTTO_TICKET
로또 번호 입력 시 구분자: LOTTO_NUMBERS_STRING_DELIMITER_REGEX

Commit:
chore: 메서드 이름 변경

Commit:
feat: 당첨 통계 제목을 출력한다

Commit:
feat: LottoTicketsBucket 클래스 addLottoTicket 메서드 테스트 코드 작성

Commit:
feat: 구매한 로또와 당첨 번호 사이에 일치하는 번호 개수 확인하는 메서드 구현

LottoTicket 클래스 sameNumbersCount 메서드 구현

해당 메서드에 대한 단위 테스트 작성

Commit:
feat: 구매한 모든 로또를 놓고 당첨 번호와 일치하는 개수를 정리해서 저장하기

LottoTicketsBucket 클래스 sameNumberCountOfAllLottoTickets 메서드 작성

해당 메서드에 대한 단위 테스트 코드 작성

Commit:
refactor: 로또 한 장 가격 이름있는 상수 별도 파일로 분리

LottoConstant 클래스 PRICE_OF_SINGLE_LOTTO_TICKET 상수로 분리

Commit:
feat: 당첨금 수익 총 합 구하는 메서드 구현

LottoTicketsBucket 클래스 sumProfit 메서드 구현

단위 테스트 작성

Commit:
feat: 수익률 계산하는 메서드 구현

MoneyToBuy 클래스 profitRatio 메서드 내부 로직 구현

단위 테스트 작성

Commit:
feat: 당첨 통계 내용 출력하기

Commit:
feat: 로또 구매를 통해서 얻은 총 수익률을 출력한다

Commit:
feat: LottoController 클래스 run 메서드 완성하기, MainApplication 구현하기

사용자가 실행할 수 있는 프로그램으로 완성

Commit:
chore: 이름있는 상수 영어 문법에 맞춰서 이름 변경

seokhongkim commented on docs/LOTTO.md:
```diff
@@ -0,0 +1,27 @@
+# 🚀 3단계 - 로또(자동)
```
요구사항 잘 정리해주셨습니다. 👍

seokhongkim commented on docs/LOTTO.md:
```diff
@@ -0,0 +1,27 @@
+    - 4개 일치하면 당첨금은 50000 원이다.
+    - 5개 일치하면 당첨금은 150000 원이다.
+    - 6개 일치하면 당첨금은 2000000000 원이다.
+- 번호 일치 개수 멸로 로또 장 수를 계산해서 최종적으로 수익률을 0 ~ 1 사이의 실수로 나타난다.
```
수익률은 1을 넘어갈 수 있을 것 같아요!

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,64 @@
+
+public class LottoController {
+    public void run() {
+        // * 사용자가 구입 금액을 입력한다.
```
주석 관련해서 클린 코드의 주석 부분을 읽어보시면 도움이 될 것 같습니다. :)

OptimistLabyrinth replied:
주석을 되도록 사용하지 말라는 의미시죠??

seokhongkim replied:
이 내용은 주석을 일부 정리했으면 하는 맥락에서 말씀드린 것이 맞습니다.

한편으로 책을 통해 어떤 주석을 남기고, 어떤 주석을 남기지 않는 것이 좋을지에 대해 고민해보셨으면 좋겠다는 의미도 있었습니다. :)

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.List;
+
+public class LottoController {
+    public void run() {
```
메소드가 20라인을 넘어가네요.
프로그래밍 요구사항을 지켜보시면 좋겠습니다.
> 함수(또는 메소드)의 길이가 15라인을 넘어가지 않도록 구현한다.

OptimistLabyrinth replied:
✅ 예, 이 부분은 개선할께요

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,64 @@
+        // * 사용자가 구입 금액을 입력한다.
+        // * 입력받은 구입 금액을 저장한다.
+        // * 거스름돈은 제거한다.
+        DemandMoneyToBuyInputPrinter.print();
```
컨트롤러가 금액을 입력받기 위해서 `DemandMoneyToBuyInputPrinter` 를 이용해 입력을 요청하고, `MoneyToBuyAcceptor` 를 이용해 값을 가져와야 한다는 것을 알아야 하네요.

컨트롤러에게 중요한 건 `구매 금액을 입력받는다` 일 것이고, `구매 금액 입력을 위한 문구를 출력하는 것` 은 `MoneyToBuyAcceptor` 의 관심사여야 하지 않을까요? :)

OptimistLabyrinth replied:
관계성이 높은 기능은 서로 묶어서 호출하도록 고쳐볼께요

seokhongkim commented on src/main/java/lotto/model/money/to/buy/MoneyToBuy.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * 로또 구매를 위해 입력한 금액을 저장하는 객체
+ */
+public class MoneyToBuy {
```
구입 금액을 값 객체로 관리하신 점 좋습니다. 👍 
로또는 1000원 단위로만 구매가 가능하니, 1000원으로 나누어 떨어지지 않는다면 예외가 발생해야 할 것 같아요!
현재 로직상으로는 1500원을 입력했을 때 1000원어치만 구매가 되는데, 수익률은 1500원을 기준으로 계산이 될 것 같습니다.

OptimistLabyrinth replied:
구현하면서 저도 그 생각을 했었습니다.
대신에 저는 거스름돈을 버리면 된다고 생각해서 현재 파일의 26 라인처럼 구현했습니다.

딱 나누어 떨어지지 않을 때 예외 발생하도록 수정할까요, 아니면 제가 한 방식을 유지해도 괜찮을까요??

seokhongkim replied:
저는 도메인의 규칙 상 1000원 단위로만 돈을 입금받을 수 있을 것이라고 생각하여 나누어 떨어지지 않으면 예외가 발생하는 것이 좋다고 생각합니다.
예외 처리를 하지 않는 경우 설명이 필요한 코드가 될 것 같습니다.

OptimistLabyrinth replied:
생각해보니까 그렇네요. 그게 더 좋은거 같아요.

seokhongkim commented on src/test/java/lotto/model/money/to/buy/MoneyToBuyTest.java:
```diff
@@ -0,0 +1,115 @@
+            super(input);
+        }
+
+        public boolean isMoneySameWith(int money) {
```
이 메소드에 대한 테스트는 `MoneyToBuy` 에 대한 테스트가 아닌 것 같아요.
테스트를 위한 클래스를 만든 후, 새로 만든 클래스에 대한 테스트만 작성했다면, 이는 프로덕션 코드에 대한 테스트가 아니니 삭제해도 괜찮을 것 같습니다. :)

OptimistLabyrinth replied:
이 부분에 대해서는 제가 어떤식으로 생각했는지 간단하게 말씀을 드려보겠습니다.

테스트 코드에서 MoneyToBuy 클래스를 상속해서  MoneyToBuyForTest 클래스를 만든 이유는
MoneyToBuy 클래스에 프로덕션에서 사용하지 않는 메서드를 추가하지 않고
MoneyToBuy 클래스의 멤버 변수가 원하는 상태에 있는지 테스트 하고 싶었어요.
그래서 원래 
private int money 였던걸 
protected int money 로 바꿨습니다.
그리고 나서 테스트 할때 MoneyToBuyForTest 클래스의 isMoneySameWith 메서드를 통해서 테스트 하면
MoneyToBuy 클래스의 money 변수의 유효성을 검증하게 된다고 생각했어요.

테스트 코드 작성 시에 이런식으로 하는 경우는 잘 없나요??
또는 이런 방식은 오류를 불러올 수 있는 생각일까요??

OptimistLabyrinth replied:
다시 보니까 이 부분은 MoneyToBuy 클래스 생성자 맨 마지막 줄(26번 라인)을 테스트하려고 만든 테스트 케이스 였네요.
생성자 테스트를 위한 테스트 스윗(test suite) 안에 있었으면
저의 의도를 명확하게 드러낼 수 있었을 것 같다는 생각이 듭니다.

이 부분은 삭제할까요, 아니면 MoneyToBuy 생성자에 대한 ```@Nested``` 테스트 스윗으로 옮길까요??

seokhongkim replied:
1000원 미만 단위의 금액을 버림 처리하고, 금액이 잘 계산되는지 검증하기 위한 테스트가 필요하군요.
이 부분은 저는 금액을 1000원 단위로 입력하지 않으면 예외가 발생되게 하고, 예외를 검증하는 것이 더 좋다고 생각합니다.

현재 로직을 유지한다면 필요한 테스트 코드라고 생각합니다.
테스트를 위해 프로덕션 코드를 변경하는 것은 좋지 않다고 생각하지만, 이 경우에는 getter 를 사용한 검증이 낫다고 생각합니다.
테스트를 위해서만 추가된 클래스의 로직은 프로덕션 코드의 로직과 달라질 수 있기에(예를 들어, isMoneySameWith 의 로직에 문제가 있을 수 있겠죠), 테스트를 위해 추가된 클래스의 로직을 통해 검증하는 것은 지양하는게 좋다고 생각합니다.

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoTicketsBucket.java:
```diff
@@ -0,0 +1,64 @@
+        this(new ArrayList<>(howManyTickets));
+    }
+
+    public LottoTicketsBucket(List<LottoTicket> lottoTickets) {
```
일부 생성자는 테스트를 위해서 추가된 생성자로 보이네요.
테스트를 위해 프로덕션 코드가 변경되어야 한다면, 이는 프로덕션 코드가 테스트에 의존하고 있는 것일수도 있습니다.
테스트만을 위한 코드를 제거하고, 프로덕션 코드를 기준으로 테스트를 해보시면 좋을 것 같아요!

OptimistLabyrinth replied:
저도 처음에는 그렇게 하려고 했는데
제가 테스트 하려는 객체의 상태에 딱 맞게 만들기 위해서 실행하는
BeforeEach 로직에 해당하는 setup 이 너무 복잡해져서 테스트 코드 작성이 너무 힘겹더라구요.

객체를 제가 테스트하려고 하는 상태로 좀 더 빠르게 만들기 위해서
생성자를 추가로 구현해서 사용했어요.

이런 경우에도 불구하고 석홍님은 정확하게 프로덕션에서 활용하는 코드만 사용해서 테스트하는게 더 나은 코드라도 생각하시나요??

seokhongkim replied:
테스트 코드 작성이 어렵다면, 이를 프로덕션 코드의 설계를 고민해봐야하는 신호로 생각해보셔도 좋겠습니다. :)
설계를 변경하기가 어렵거나, 비용이 너무 크다면 불가피하게 프로덕션 코드에 테스트에 필요한 일부 코드를 추가할 수 있겠습니다. 그럼에도 저는 기본적으로 프로덕션 코드를 테스트를 위해 변경하는 것은 지양해야 한다고 생각합니다.

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,64 @@
+        final LottoTicketsBucket lottoTicketsBucket = new LottoTicketsBucket(numberOfAffordableLotto);
+        while (0 < numberOfAffordableLotto--) {
+            final List<Integer> candidates = intsFromOneToFortyFive();
+            final LottoNumberGenerator lottoNumberGenerator = new LottoNumberGenerator(candidates);
```
매번 새로운 값을 저장하는 새로운 객체를 만들고, `Random` 위치에 있는 특정 값을 꺼내서 반환하는 형태이다보니 객체 생성이 많이 발생할 수 있겠네요.
로또에서 생성할 수 있는 번호 목록을 만들어 놓고, `Collections.shuffle` 을 이용해서 섞은 후 앞에서 N개의 번호만 반환하면 임의의 수를 매번 반환해줄 수 있을 것 같아요!
미션에 있는 힌트를 참고해보시면 좋겠습니다. :)

> https://edu.nextstep.camp/s/urD8knTu/ls/DDyZo2rO
로또 자동 생성은 Collections.shuffle() 메소드 활용한다.

OptimistLabyrinth replied:
좋은 아이디어네요 ㅎㅎ 그렇게 수정해볼께요

seokhongkim commented on src/main/java/lotto/view/LottoTicketPrinter.java:
```diff
@@ -0,0 +1,12 @@
+ */
+public class LottoTicketPrinter {
+    public static void print(LottoTicket lottoTicket) {
+        System.out.println(lottoTicket);
```
발급한 로또 번호를 출력할 때 정렬해서 출력해야 하는데, 이에 대한 로직은 어디에 있을까요?

OptimistLabyrinth replied:
요구사항을 제가 착각한 것 같아요 😆
정렬하는 로직도 추가할께요

seokhongkim commented on src/main/java/lotto/constant/numbers/LottoConstant.java:
```diff
@@ -0,0 +1,13 @@
+package lotto.constant.numbers;
+
+public class LottoConstant {
```
상수 클래스가 유용한 경우도 있지만 주의해서 사용이 필요하다고 생각합니다.
1. 로또 번호의 최솟값과 최댓값, 로또 번호의 개수 등은 도메인의 규칙인데, `LottoConstant` 에서 관리되면 도메인이 스스로 알고 있어야 하고 관리해야 할 정책을 찾기 위해 한 단계를 더 찾아야 합니다. 이는 도메인의 규칙을 찾기 어렵게 만들고, 코드 수정도 어렵게 만드는 경향이 있습니다.
2. 공통 상수로 관리하는 경우 자칫 같은 값을 가진 다른 목적의 상수가 하나로 정의되는 경우가 생길 수 있습니다.
3. 여러 상수가 정의되었을 때 맥락에 따라 구분을 해야 한 눈에 들어오는데, 구분이 어려운 경우가 생기면 상수가 정의된 코드가 한 눈에 안 들어오기도 합니다.
4. 상수 클래스는 값을 `public` 으로 선언하게 되는데, 이는 특정 객체가 가지고 있어야 할 책임을 다른 곳에서도 쉽게 접근하게 만듭니다. 예를 들어, 로또는 6개의 번호로 구성되어야 한다는 것은 도메인 규칙이고 다른 곳에서는 임의로 접근하지 않아야 합니다. 하지만 다른 곳에서도 6이라는 값이 필요할 때 쉽게 접근할 수 있고, 이는 도메인 규칙이 외부에서 쉽게 사용되게 만들며 강한 결합을 발생시킬 수 있습니다.

OptimistLabyrinth replied:
그렇군요.
나름 여러 군데에서 사용한다고 생각해서 공통으로 분리한거긴 한데
cohesion? 응집성? 이라는 관점에서 생각하면 이렇게 두는게 안 좋을 수도 있겠네요

OptimistLabyrinth replied:
상수를 static 으로 선언하는 것도 문제가 될까요??

seokhongkim replied:
상수를 static 으로 관리하시는 것은 좋습니다. 👍

seokhongkim commented on src/main/java/lotto/constant/utils/StringUtils.java:
```diff
@@ -0,0 +1,7 @@
+package lotto.constant.utils;
+
+public class StringUtils {
```
유틸리티 클래스 만들어서 사용하시는 것 좋습니다. 👍

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Random;
+
+/**
+ * 1 이상 45 이하의 번호 중에서 무작위 번호를 하나 선택하는 객체이다.
```
구현만 봤을 때 1 이상 45 이하라는 것은 보장되지 않는 것 같아요.

OptimistLabyrinth replied:
이거는 주석이 개발자를 헷갈리게 한 부분이라고 봐야겠군요.
LottoNumberGenerator 클래스, LottoTicket 클래스, LottoTicketsBucket 클래스를 구현하다보니
테스트 코드 작성이나 이런 부분까지 생각하다보니
LottoNumberGenerator 에서 사용하는 "번호 생성 후보군" 을 외부에서 주입받는게 더 낫다고 판단하게 되었어요.

private final List<Integer> lottoNumberCandidates 라는
멤버 변수 자체도 LottoNumberGenerator 객체가 스스로 생성해서 갖고 있는게 나을까요???

seokhongkim replied:
`LottoNumberGenerator` 는 정해진 범위의 `LottoNumber` 중에서만 난수를 반환(혹은 6개의 번호를 반환할 수도 있겠네요.)해야 할 것 같아요.
이를 외부에서 주입할 수 있도록 하는 것은 `LottoNumberGenerator` 가 지켜야 할 규칙을 어렵게 만드는 것은 아닐까요?
`외부에서 주입된 목록에서 랜덤한 값을 1개씩 꺼내주는 클래스`라면, `LottoNumberGenerator` 가 아닌 `RandomNumberPicker` 와 같은 네이밍이 되어야 하지 않을까요? :)

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoTicket.java:
```diff
@@ -0,0 +1,48 @@
+ * 자동 생성한 6 개의 로또 번호 한 줄을 저장하고 관리하는 객체이다.
+ */
+public class LottoTicket {
+    protected final List<Integer> numbers;
```
로또 번호를 관리하기 위한 객체를 구현해보시면 좋겠습니다.
1에서 45 사이의 값만 가지고 있어야 하는데, `Integer` 를 사용하면 이에 대한 보장이 어려울 것 같아요.

OptimistLabyrinth replied:
아이디어 제안 고맙습니다.
그런 객체를 만들면 WinningNumbers 쪽에서도 동시에 사용할 수 있을 것 같네요 😄👍👍

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoTicket.java:
```diff
@@ -0,0 +1,48 @@
+        }
+    }
+
+    public LottoTicket(List<Integer> numbers) {
```
로또는 6개의 숫자로 구성되어 있어야 하고 중복이 없어야 하는데, 이에 대한 밸리데이션을 추가해보는 건 어떨까요?

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoTicket.java:
```diff
@@ -0,0 +1,48 @@
+    public LottoTicket(LottoNumberGenerator lottoNumberGenerator) {
+        numbers = new ArrayList<>(LottoConstant.COUNT_OF_NUMBERS_IN_SINGLE_LOTTO_TICKET);
+        for (int i = 0; i < LottoConstant.COUNT_OF_NUMBERS_IN_SINGLE_LOTTO_TICKET; ++i) {
+            final int generatedRandomNumber = lottoNumberGenerator.generate();
```
`LottoNumberGenerator` 가 `generate` 를 요청할 때마다 중복 없이 값을 반환해준다는 사실을 알고 있고, `COUNT_OF_NUMBERS_IN_SINGLE_LOTTO_TICKET` 횟수만큼 호출해도 안전하다는 사실도 알고 있어야 하네요.
`LottoTicket` 이 `LottoNumberGenerator` 의 구현을 자세히 알고 사용하는 것은 아닐까요?

OptimistLabyrinth replied:
저도 그 부분이 마음에 걸리긴 하는데
테스트 작성이 편해지려면 계속 잘게 분리하는게 편한 것 같아서 이런 모양이 되어버렸어요.

OptimistLabyrinth replied:
혹시 어떻게 수정하면 좋을지
생각나는 간단하게 방법을 간단하게 소개해주실 수 있나요??

seokhongkim replied:
현재 `LottoTicket` 이 `LottoNumberGenerator` 에 의존하고 있는데, 이 의존 방향을 반대로 구현해보셔도 좋겠습니다.
`LottoNumberGenerator` 대신 `AutoLottoGenerator`, `AutoLottoIssuer` 와 같은 클래스를 만든다고 상상해보면, 해당 클래스는 `자동으로 6개의 번호를 추출하여 Lotto를 생성`하는 역할을 할 수 있을 것 같습니다. :)

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoTicketsBucket.java:
```diff
@@ -0,0 +1,64 @@
+        return lottoSameNumberCount;
+    }
+
+    public int sameThreeNumbersCount() {
```
로또가 확장되어 번호가 20개 일치하는 케이스까지 늘어난다면 구현이 어떻게 될까요? :)
비슷한 코드가 반복되면 호출하는 과정에서 실수할 수 있고, 변경이 어려워집니다.
`List` 또는 `Array` 가 아닌 `등수와 개수` 의 관계를 쉽게 관리할 수 있는 자료 구조를 사용해보셔도 좋겠습니다.

hint: `Map`

OptimistLabyrinth replied:
아 ㅋㅋㅋㅋㅋㅋㅋ 그렇네요. 제안 내용대로 수정해볼께요

seokhongkim commented on src/main/java/lotto/model/winning/numbers/WinningNumberEach.java:
```diff
@@ -0,0 +1,40 @@
+        }
+        final String trimmedToken = token.trim();
+        int tokenValue;
+        try {
```
예외로 발생시키는 일부 규칙이 `로또에 사용되는 숫자` 가 지켜야 할 규칙으로 보이네요!

seokhongkim commented on src/main/java/lotto/model/winning/numbers/WinningNumbers.java:
```diff
@@ -0,0 +1,54 @@
+public class WinningNumbers {
+    private final List<WinningNumberEach> numbers;
+
+    public WinningNumbers(String input) {
```
문자를 입력 받아서 숫자로 컨버팅하는 것은 입력에서 관리해야 할 로직은 아닐까요?
컨트롤러에 남긴 코멘트처럼 로또라는 코드를 그대로 남기고 콘솔에서 웹 환경으로 변경할 때, 어느 코드를 변경해야 하는지 한번 고민해보셔도 도움이 될 것 같습니다.

OptimistLabyrinth replied:
문자를 입력받아서 숫자로 바꾸는 것은
Model 이 아니고 Controller 에서 해야하는 역할이라고 생각하시는 것 같습니다.

제가 석홍님의 생각을 제대로 이해한게 맞나요???

seokhongkim replied:
저는 Controller 또는 InputView(Acceptor 로 구현하신 클래스) 에서 해야 할 역할이라고 생각합니다.
이를 통해 도메인의 재사용성이 증대되고, UI에 대한 의존성도 제거할 수 있을 것 같습니다. `WinningNumber` 는 `WinningNumber` 가 알아야 하는 주요 로직에 집중하면 좋겠습니다.

OptimistLabyrinth replied:
무슨 의미인지 알 것 같네요. 수정해볼께요 😄👍👍

seokhongkim commented on src/main/java/lotto/view/LottoWinningStatisticsPrinter.java:
```diff
@@ -0,0 +1,21 @@
+ * 3 개, 4 개, 5 개, 6 개 일치하는 개수를 한 줄 씩 출력한다.
+ */
+public class LottoWinningStatisticsPrinter {
+    private static final String THREE_DIGITS_MATCH_MESSAGE = "3개 일치 (5000원)- ";
```
등수를 관리하는 객체를 활용해보면 어떨까요?
1등은 6개의 번호가 일치해야 하고, 20억의 상금을 가지고 있네요.

hint: `enum`

OptimistLabyrinth replied:
한번 고민해볼께요

seokhongkim commented on src/test/java/lotto/model/lotto/ticket/LottoTicketTest.java:
```diff
@@ -0,0 +1,126 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class LottoTicketTest {
```
단위 코드 테스트를 꼼꼼하게 작성해주셨네요! 👍

seokhongkim commented on src/main/java/lotto/view/BlankLinePrinter.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * 빈 줄을 한 줄 출력한다.
+ */
+public class BlankLinePrinter {
```
하나하나의 출력을 각 클래스로 분리하셨네요. 이렇게 하니 각 클래스가 무엇을 출력하는지 잘 들어오는 장점이 있네요.

현재 구조와 관련하여 두 가지 생각이 들었습니다.
1. 컨트롤러가 모든 구현체를 알아야 하네요. 한 클래스에 여러 메소드를 넣고 무엇을 출력하는지를 메소드의 이름으로 구분할지, 클래스의 이름으로 구분할지에 대한 선택인 것 같습니다. 구현을 직접 하지 않은 개발자 입장에서 사용할 때, 클래스 하나에 메소드가 있으면 자동완성이 용이하지만 N개의 클래스가 있다면 무엇이 있는지 직접 찾아야 할 것 같기도 합니다.
2. 출력하는 내용이 너무 구체적인 단위로 나뉘어 있어서, 컨트롤러가 결국 출력을 하는 순서에 맞춰 메소드를 다 호출해줘야 하는 것으로 보입니다. 예를 들어, 구매한 로또 결과를 출력할 때 컨트롤러는 "구매 결과를 출력해줘"가 관심사일테고, 출력은 "N장을 구매했고 각 내용을 보여줘"가 관심사가 될 수 있을텐데, 현재는 컨트롤러가 출력 내용에 대해서도 모두 알아야 하는 것으로 보이네요. 결과를 출력하는 것도 모두 컨트롤러가 자세히 알아야 하는 것 같습니다.

OptimistLabyrinth replied:
메시지 출력하는 로직을 사용자 입력을 받는 객체로 이동하라는 말씀이시죠??

OptimistLabyrinth replied:
이건 좀 어려운 것 같습니다.

이 컨트롤러 로직도 테스트 코드 작성하기 쉽도록 잘 분리하고 싶은데
어떻게 해야할지 잘 모르겠네요.
내부적으로 사용하는 모든 의존성을 외부에서 주입받도록 구현해야 하는걸까요

석홍님은 어떻게 하실 것 같으신가요??

OptimistLabyrinth replied:
유지보수, 가독성 증대 측면에서 
별도 클래스로 구분하는 것보다
하나의 클래스 내에 여러 개의 메서드로 구분하는 게

더 좋은 것 같다는 말씀이신가요??

seokhongkim replied:
> 이 컨트롤러 로직도 테스트 코드 작성하기 쉽도록 잘 분리하고 싶은데
어떻게 해야할지 잘 모르겠네요.
내부적으로 사용하는 모든 의존성을 외부에서 주입받도록 구현해야 하는걸까요
석홍님은 어떻게 하실 것 같으신가요??

컨트롤러가 의존하는 클래스가 많을수록 외부에서 주입받기 어려울 것 같습니다.
컨트롤러가 입력값을 읽어오는 객체, 결과를 출력하는 객체만 외부에서 받아오면 어떨까요? 그럼 컨트롤러 테스트를 위해 사용자가 입력하는 값은 입력값을 읽어오는 객체의 행동을 변경해서 테스트할 수 있을 것 같습니다. 예를 들어, 사용자의 금액 입력을 요청하면 5000이라는 값을 반환하도록 객체의 행위를 조작할 수 있겠죠.
출력에 대한 검증도 출력하는 객체의 메소드를 조작해서 검증이 가능할 것 같습니다.
도메인의 로직은 입력에 따라 정상적으로 실행될 것이고, 출력 메소드를 통해 값의 검증이 가능할 것 같습니다.
이 내용은 어려우시면 처음에는 컨트롤러에 대한 테스트는 작성하지 말고, 인풋과 아웃풋 처리를 위한 클래스만 하나로 변경해서 구현해보셔도 좋겠습니다.
그리고 static method를 많이 사용할수록 강한 결합이 되고 주입이 어려워져 테스트 작성이 어렵습니다. 컨트롤러를 아래 구조로 생성하는 것으로 시작해보시면 도움이 될 것 같습니다.
```java
public class LottoController {
    private final InputView inputView;
    private final OuputView outputView;

    public LottoController(InputView inputView, OuputView outputView) {
        this.inputView = inputView;
        this.outputView = outputView;
    }
}
```

> 유지보수, 가독성 증대 측면에서
별도 클래스로 구분하는 것보다
하나의 클래스 내에 여러 개의 메서드로 구분하는 게
더 좋은 것 같다는 말씀이신가요??

네, 저는 지금은 그렇게 생각합니다.
기성님께서 다르게 생각하실 수 있고, 그 방향이 더 좋을 수 있기에 고민해보시고 함께 이야기 나눠봐도 좋겠습니다. :)

OptimistLabyrinth replied:
오 이거 괜찮은데요??? ㅋㅋㅋㅋ 왜 이걸 진작 생각하지 못했을까요
수정해볼께요

seokhongkim reviewed (changes requested):
기성님, 3단계 미션 잘 구현해주셨습니다. 👍 
역할을 분리하기 위해 노력하신 것 같고, 단위 테스트도 꼼꼼하게 작성해주셨네요!

조금 더 고민해보시면 도움이 될 것 같은 내용을 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

Commit:
docs: 문서 오류 수정

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1008780315](https://github.com/next-step/java-lotto-pro/pull/802/files#r1008780315)

OptimistLabyrinth reviewed (commented):
매번 정성스럽게 리뷰해주셔서 감사해요 👍👍
달아주신 코멘트에 대해서 추가로 궁금한 부분 답변 달았어요.

읽어보시고 답변 주시면 감사하겠습니다.

Commit:
refactor: 불필요한 주석 제거, 메서드 길이 10줄 이내로 만들기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008780406](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008780406)
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008780509](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008780509)

Commit:
refactor: 입력 요구를 위한 출력을 stdin 에서 입력값 받아서 저장하는 클래스로 이동

higher cohesion, the better

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008780658](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008780658)

Commit:
refactor: 반드시 필요하지 않은 주석은 전부 삭제

주석은 제대로 유지보수 되지 못할 가능성이 높다
그러므로 주석은 최대한 지양하는게 좋다

Commit:
feat: 한 장의 로또를 출력할 때 6 개의 숫자 오름차순 정렬

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008781384](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008781384)

Commit:
feat: 로또 번호를 저장하고 관리하기 위한 클래스 만들기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008781707](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1008781707)

OptimistLabyrinth reviewed (commented):
코멘트 추가했습니다 ㅎㅎ

OptimistLabyrinth commented:
프로젝트를 괜찮은 구조로 구성하는게 쉽지가 않네요.
다 지우고 다시 처음부터 시작할까요??

seokhongkim commented:
> 프로젝트를 괜찮은 구조로 구성하는게 쉽지가 않네요.
다 지우고 다시 처음부터 시작할까요??

제가 이번에 피드백을 많이 남겨서 기성님께서 당황스러우셨을 것 같기도 하고, 좋은 구조에 대해 고민이 되셨을 것 같기도 합니다.
저는 처음부터 다시 시작하시는 것보다 현재 구조를 조금씩 바꿔보는 경험을 하시는 것도 도움이 될 것 같아요. 이유는 아래와 같습니다.
1. 먼저, 제가 드린 리뷰 요청 내용이 정답은 아닙니다. 그리고 기성님께서 처음에 이렇게 개발하신 이유가 있을 것이라고 생각합니다. 제 리뷰에 맞춰 기성님의 코드를 고치는 것이 아니라, 제가 말씀드린 내용 중 일부는 수용하실수도 있을 것 같고, 기성님이 생각하시기에 적절하지 않다면 기성님의 생각을 말씀해주셔도 좋겠습니다.
2. 제 경험으로는 개발하는 과정에서 코드를 처음부터 만드는 일보다, 이미 만들어진 코드를 요구사항에 맞춰 변경하고 리팩토링하는 일이 더 많은 것 같습니다. 그래서, 이번 미션에서도 제 리뷰에 맞춰 처음부터 만드시거나, 새로운 요구사항에 맞춰 처음부터 개발하시는 것보다는 현재 만드신 코드를 리팩토링하면서 적정한 지점을 찾아가는 것이 기성님의 학습과 성장에 더 도움이 될 것이라고 생각합니다.

seokhongkim reviewed (changes requested):
기성님, 코멘트 남겨주신 부분에 대해 답변 남겼습니다. :)
확인 부탁드립니다.

Commit:
feat: LottoNumber 클래스 수정

LottoNumber 클래스는 LottoNumber, WinningNumbers 두 개의 클래스에서 사용한다

Commit:
feat: LottoTicket 클래스 수정

Commit:
feat: MoneyToBuyAcceptor 클래스 수정

Commit:
feat: LottoNumberGenerator 클래스 제거

Commit:
feat: LottoTicketsBucket 클래스 수정

Commit:
feat: MoneyToBuy 클래스 수정

Commit:
feat: WinningNumbers 클래스 수정

Commit:
feat: WinningNumberEach 클래스 제거

Commit:
feat: LottoConstant 클래스 제거

Commit:
feat: MoneyToBuy 클래스 canBuyMoreLotto, buyOneLotto 메서드 추가

구매 금액을 가지고 가능한 만큼 로또를 구매하는 로직을 구현할 때
사용하는 메서드이다

MoneyToBuy 클래스에 대한 단위 테스트 추가

Commit:
feat: 잘못된 import 수정

Commit:
feat: LottoTicketsBucket 클래스 addLottoTicket 메서드 추가

Commit:
feat: LottoTicket, LottoNumber 클레스에서 toString 메서드 오버라이드

Commit:
feat: LottoResult 클래스 추가

OptimistLabyrinth reviewed (commented):
어제는 제가 늦게까지 처리해야하는 일이 있어서 답변이 늦었습니다.
피드백 주신 부분 중점으로 해서 수정해볼께요

Commit:
feat: LottoController 로직 되살리기

잠시 주석처리 했던 LottoController 코드를
다시 원래대로 주석처리 해제한다

Commit:
feat: LottoTicket 클래스에서 무작위 숫자 만들어내는 로직과의 결합도 낮추기

Commit:
feat: LottoNumberGenerator 로직 오류 수정

Commit:
test: LottoTicket 클래스에 대한 테스트에서 딱 필요한 것만 남기기

Commit:
chore: 멤버 변수 이름 변경

Commit:
feat: 로또 결과 계산 로직 LottoResult 맞춰서 개선

List<T>.contains() 메서드 사용해서 확인할 때
구현 로직대로 동작하도록 만들기 위해서
LottoNumber 클래스에서 equals, hashCode 메서드 오버라이드

LottoTicketsBucket 단위 테스트 추가
LottoTicket 단위 테스트 추가

Commit:
feat: 총 수익률 구하는 로직 구현

LottoResult 클래스 profitRatio 메서드 구현
MoneyToBuy 클래스 profitRatio 메서드 구현

단위 테스트 작성

Commit:
feat: 호출의 대상 아니고 테스트 대상인 메서드는 public 대신에 protected

Commit:
feat: 당첨 통계 출력 로직 구현

LottoController 클래스 displayLottoResult 메서드

Commit:
feat: (피드백 반영) 문자를 입력받아서 숫자로 컨버팅하는 로직 controller 로 이동

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1008881038](https://github.com/next-step/java-lotto-pro/pull/802/files#r1008881038)

Commit:
feat: (피드백 반영) WinningNumbersConverter 클래스 WinningNumbersAcceptor 로 이동

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1008782675](https://github.com/next-step/java-lotto-pro/pull/802/files#r1008782675)

Commit:
feat: controller 패키지 내의 레이어를 IOC 사용해서 개선

STDIN Acceptor 클래스 coupling 완화

Commit:
chore: 필요없는 메서드 제거

Commit:
chore: 필요없는 import 제거

OptimistLabyrinth commented:
@seokhongkim 

쉽지 않군요 ㅎㅎ
피드백 코멘트 하나씩 보면서 최대한 수정해보려고 했습니다

seokhongkim commented on src/main/java/lotto/MainApplication.java:
```diff
@@ -0,0 +1,18 @@
+
+public class MainApplication {
+    public static void main(String[] args) {
+        final int[][] prizes = new int[][]{{3, 5000}, {4, 50000}, {5, 150000}, {6, 2000000000}};
```
일치하는 개수와 상금은 로또의 규칙이므로 도메인에서 enum 을 활용해서 관리해보셔도 좋겠습니다. :)

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,83 @@
+        final MoneyToBuy moneyToBuy = userInputMoneyToBuy();
+        int numberOfAffordableLotto = moneyToBuy.affordableTicketCount();
+        final LottoTicketsBucket lottoTicketsBucket = new LottoTicketsBucket(numberOfAffordableLotto);
+        while (moneyToBuy.canBuyMoreLotto()) {
```
`MoneyToBuy` 를 이용해서 로또를 몇 장이나 구매했는지 관리하셨군요.
클래스가 하는 역할이 늘어나고, 이름으로 설명하기 어려운 역할인 것 같아요.
1. 로또를 구매할 돈을 관리한다.
2. 구매한 로또의 수량을 관리한다.

돈으로부터 몇 장이나 구매할 수 있는지는 `MoneyToBuy` 가 알려주는 것이 괜찮다고 생각하지만, 실제로 몇 장이나 구매했는지가 `MoneyToBuy` 의 관심사인지는 고민해보시면 좋겠습니다. :)
로직상 `LottoTicketsBucket` 도 몇 장을 구매했는지 알고 있을 것 같네요!

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,83 @@
+        return moneyToBuy;
+    }
+
+    private void numberOfAffordableLottoLeft(LottoTicketsBucket lottoTicketsBucket) {
```
로또를 구매하는 행위를 하는데, 메소드 명은 이를 잘 설명해주지 못하는 것 같아요.
`로또를 구매한다`는 행위를 설명해보면 어떨까요?

seokhongkim replied:
`LottoTicketsBucket` 의 역할이 구매한 로또를 담고 있는 것이라면, `로또를 구매한다` 또는 `로또를 발급한다` 는 메소드를 통해 발급한 `Lotto`를 반환 받아서 메소드 외부에서 결과를 출력하고, `LottoTicketsBucket` 에 담는 것이 더 이해하기 쉬울 것 같네요.

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    private LottoResult displayLottoStatistics(LottoTicketsBucket lottoTicketsBucket, WinningNumbers winningNumbers) {
+        final Map<Integer, Integer> prizeMoney = new HashMap<>();
```
`MainApplication` 에 남긴 코멘트처럼 prize 는 enum 으로 관리해보시는 게 좋겠습니다.
또한 `calculateNumbersMatchCount` 에게 `Map` 을 이용해서 요청을 보내는데, `Map` 구조는 key 와 value 를 알아야 하므로 저장하는 값에 대해 자세히 알아야 하고, 특히나 key 와 value 가 같은 타입인 경우 더 어려울 수 있습니다. :)

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,83 @@
+        displayProfitRatio(lottoResult, moneyToBuy);
+    }
+
+    private LottoResult displayLottoStatistics(LottoTicketsBucket lottoTicketsBucket, WinningNumbers winningNumbers) {
```
메소드의 이름은 `display` 하는 것인데, 실제로는 `LottoResult` 를 계산하는 행위와 결과를 출력하는 행위를 모두 하고 있네요.
메소드를 나눠보셔도 좋겠습니다.

OptimistLabyrinth replied:
수정 완료했습니다.

여기 외에 다른 곳도 필요한 부분을 찾아서 메서드 분리했습니다.

seokhongkim commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,83 @@
+        for (int[] prize : prizes) {
+            prizeMoney.put(prize[0], prize[1]);
+        }
+        final Map<Integer, Integer> countsOfNumbersMatch = lottoTicketsBucket.calculateNumbersMatchCount(prizeMoney,
```
`calculateNumbersMatchCount` 는 구현을 설명하는 이름으로 보입니다.
로또와 우승 번호를 비교하며 `당첨 여부를 확인`하는 것이 해당 메소드의 기능 아닐까요? :)

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoNumber.java:
```diff
@@ -0,0 +1,57 @@
+            throw new NumberFormatException("개별 당첨 번호는 반드시 숫자여야 합니다.");
+        }
+        if (lessThanOneOrGreaterThanFortyFive(tokenValue)) {
+            throw new NumberFormatException("개별 당첨 번호는 1 이상 45 이하의 숫자 중에 하나여야 합니다.");
```
메소드 전체적으로 `NumberFormatException` 보다는 `IllegalArgumentException` 이 `로또 번호`의 입장에서 `인자가 잘못되었음`을 더 잘 설명해줄 수 있는 예외 같습니다.

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoNumber.java:
```diff
@@ -0,0 +1,57 @@
+        this.value = value;
+    }
+
+    private boolean lessThanOneOrGreaterThanFortyFive(int tokenValue) {
```
메소드의 구현을 설명하는 이름보다 행위를 설명해주는 이름이 좋습니다.
예를 들어, 숫자의 범위가 1-50으로 늘어나면 이 메소드는 `lessThanOneOrGreaterThanFifty` 로 이름을 함께 변경해줘야 합니다.
반면, `유효한 로또 숫자인가` 혹은 `숫자가 유효한지 검증한다` 와 같이 `isValidLottoNumber`, `validateLottoNumber (예외 던지는 로직을 이동하는 경우)` 이름을 지으면 범위가 달라져도 행위가 변하지 않으므로 메소드명은 그대로 사용할 수 있습니다.

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,35 @@
+
+public class LottoNumberGenerator {
+    protected static final int NUMBER_COUNT_IN_SINGLE_LOTTO_TICKET = 6;
+    private static final int LOTTO_MINIMUM_NUMBER = 1;
```
1-45 사이라는 값을 `LottoNumber` 에서도 관리하고, `LottoNumberGenerator` 에서도 관리하게 되는데 `LottoNumberGeneartor` 가 `Integer` 리스트를 이용하여 숫자를 먼저 정한 후 `LottoNumber` 로 컨버팅하기 때문에 필요한 값으로 보입니다.

`LottoNumber` 가 생성될 수 있는 객체(1-45 사이의 값)를 미리 목록으로 가지고 있다면, `LottoNumberGenerator` 는 그 목록을 요청해서 랜덤하게 6개의 값만 꺼내주면 될 것 같아요.
예시 코드를 보면 아래와 같습니다.
```java
public class LottoNumber
  // 중략
  private static final List<LottoNumber> LOTTO_NUMBERS;

  static {
    LOTTO_NUMBERS = Collections.unmodifiableList(
      IntStream.rangeClosed(LOTTO_MIN_VALUE, LOTTO_MAX_VALUE)
        .mapToObj(LottoNumber::new)
        .collect(Collectors.toList())
    );
  }

  public LottoNumber(int lottoNumber) {
    validateLottoNumber(lottoNumber);
    LOTTO_NUMBERS.get(lottoNumber - 1);  // 리스트는 0번부터 시작하므로 1을 빼줌. 맵으로 관리하거나, 0번을 리스트에 추가하면 더 이해하기 쉽게 구현 가능.
  }

  public static List<LottoNumber> selectableLottoNumbers() {
    return new ArrayList<>(LOTTO_NUMBERS);
  }
}

public class LottoNumberGenerator {
  public List<LottoNumber> generate() {
    List<LottoNumber> selectableLottoNumbers = LottoNumber.selectableLottoNumbers();
    Collections.shuffle(selectableLottoNumbers);
    return selectableLottoNumbers.subList(0, NUMBER_COUNT_IN_SINGLE_LOTTO_TICKET);
  }
}
```

OptimistLabyrinth replied:
```java
class A {
    static {
    }
}
```

처럼 클래스 내부에 static 할 수 있는건 처음 알았네요 👍👍

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,35 @@
+    public List<LottoNumber> generate() {
+        Collections.shuffle(fullCandidateList);
+        final List<Integer> numbers = fullCandidateList.subList(0, NUMBER_COUNT_IN_SINGLE_LOTTO_TICKET);
+        Collections.sort(numbers);
```
정렬을 하는 것이 언제 필요한지 고민해보셔도 좋겠습니다. :)
`Lotto` 에게 필요한 규칙이라면, `Lotto` 생성자 내부에서 하는 것이 좋아 보입니다. 외부에서 어떻게 값을 전달할지 보장할 수 없기 때문입니다.
출력 할 때 필요한 규칙이라면 출력하는 곳에서만 정렬해서 보여주는 것도 방법일 것 같습니다.

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoTicket.java:
```diff
@@ -0,0 +1,34 @@
+public class LottoTicket {
+    protected final List<LottoNumber> lottoNumbers;
+
+    public LottoTicket(LottoNumberGenerator lottoNumberGenerator) {
```
`LottoTicket` 이 `LottoNumberGenerator` 를 전달 받아서 값을 생성해야 할 이유가 있을까요?
`LottoNumberGenerator` 가 `List<LottoNumber>` 를 생성한 후, 이 값을 `LottoTicket` 의 생성자에게 전달해서 객체를 만드는 방법은 어떨까요?

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/model/lotto/ticket/LottoTicket.java:
```diff
@@ -0,0 +1,34 @@
+        return count;
+    }
+
+    private int incrementCountIfLottoNumberMatch(int count, LottoNumber winningNumber) {
```
`count + 1` 값을 반환하기 위해 `count` 인자를 전달 받아야 하네요.
메소드가 값이 있는 경우 1을 반환해주고, `numberMatch` 에서 `count +=` 의 형태로 호출해주면 `count` 인자를 제거할 수 있을 것 같아요!

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim reviewed (changes requested):
기성님, 3단계 미션 피드백 잘 반영해주셨습니다. 👍 

3단계 미션에서 조금 더 고민해보시면 좋을 것 같은 부분이 있어 코멘트 추가로 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

Commit:
feat: 일치하는 번호 개수 enum 으로 관리하기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012566417](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012566417)
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012572459](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012572459)

Commit:
feat: MoneyToBuy 클래스에 어울리지 않은 로직 다른 클래스 이동

로또를 구매하는 책임은 LottoTicketsBucket 클래스에서 갖는다

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012569233](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012569233)

Commit:
feat: 제대로 동작하지 않는 LottoController 로직 수정

Commit:
feat: 메서드 이름과 실제 동작이 다른 경우 삭제

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1012569862](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1012569862)

Commit:
refactor: 가독성 향상을 위해서 별도 메서드로 분리

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1012569862](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1012569862)

Commit:
refactor: 새로운 로또를 구매하는 것과 방금 구매한 로또를 출력하는 메서드 분리

Commit:
feat: 메서드를 SRP 원칙에 따라서 분리하기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1012572888](https://github.com/next-step/java-lotto-pro/pull/802#discussion_r1012572888)

Commit:
feat: 유효하지 않은 로또 번호 입력값에 대한 예외를 IllegalArgumentException 타입으로 변경

< 수정 전 >
기존에 던지는 예외는 NumberFormatException

< 수정 후 >
이런 맥락에서 사용하기에 더 적절한 예외는 NumberFormatException

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012575968](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012575968)

Commit:
refactor: 메서드 이름으로 구현 내용이 아닌 역할을 설명하기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012581124](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012581124)

Commit:
refactor: 메서드 파라미터 개수 줄이기

둘 이상의 메서드에서 공통으로 사용하는 변수는
파라미터로 전달하기 보다는 멤버 변수로 바꾸기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012589040](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012589040)

Commit:
feat: 생성자에 전달하는 파라미터 변경

< 수정 전 >
LottoTicket 생성자에서 LottoNumberGenerator 객체를 받기

< 수정 후 >
LottoTicket 생성자에서
LottoNumberGenerator 객체가 생성한 List<LottoNumber> 전달 받기

< 관련 링크 >
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012588099](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012588099)

Commit:
refactor: 동일한 데이터 중복으로 관리하지 않도록 리팩터링

< 수정 전 >
로또 번호 최솟값 1, 최댓값 45 를 LottoNumberGenerator, LottoNumber 클래스
양쪽에서 관리하고 있었다

< 수정 후 >
로또 번호 후보군 리스트를 LottoNumber 클래스에서 저장하고 있다가
LottoNumberGenerator 객체가 generate 메서드 호출하는 순간
해당 후보군 리스트를 가져와서 사용한다

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012586865](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012586865)

Commit:
refactor: 로또 번호 정렬은 출력하는 순간에 하기

관련 링크
[https://github.com/next-step/java-lotto-pro/pull/802/files#r1012587586](https://github.com/next-step/java-lotto-pro/pull/802/files#r1012587586)

seokhongkim commented on src/main/java/lotto/model/money/to/buy/MoneyToBuy.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private boolean divisibleWithoutRemainder(int inputMoney) {
+        int moneyRemainderDiscarded = inputMoney / PRICE_OF_SINGLE_LOTTO_TICKET * PRICE_OF_SINGLE_LOTTO_TICKET;
```
로또 가격으로 나눈 금액의 나머지가 0원인지 검증하는 것으로 써볼 수도 있겠네요! :)
```suggestion
        return (inputMoney % PRICE_OF_SINGLE_LOTTO_TICKET) == 0;
```

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/view/NumberOfLottoTicketsPrinters.java:
```diff
@@ -0,0 +1,11 @@
+    private static final String NUMBER_OF_AFFORDABLE_LOTTO_TICKETS_MESSAGE = "개를 구매했습니다.";
+
+    public static void print(MoneyToBuy moneyToBuy) {
+//        System.out.println(moneyToBuy.affordableLottoTicketNumber() + NUMBER_OF_AFFORDABLE_LOTTO_TICKETS_MESSAGE);
```
사용하지 않는 코드는 주석으로 변경하지 말고 삭제하시는 게 좋습니다.
* 과거의 코드가 필요한 경우 git 이력을 통해 다시 찾을 수 있습니다.
* 주석 처리된 코드인 경우 의도적으로 처리한 것인지, 실수로 처리한 것인지 헷갈리는 때가 있기도 합니다. 또한 실수로 주석 처리를 푸는 경우도 발생합니다.

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/model/lotto/enums/LottoNumberMatchCount.java:
```diff
@@ -0,0 +1,18 @@
+public enum LottoNumberMatchCount {
+    THREE(3),
+    FOUR(4),
+    FIVE(4),
```
FIVE는 5개고, SIX는 6개의 값을 가질 것 같아요!

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim commented on src/main/java/lotto/MainApplication.java:
```diff
@@ -0,0 +1,26 @@
+public class MainApplication {
+    public static void main(String[] args) {
+        final Map<LottoNumberMatchCount, Integer> prizeMoney = new HashMap<>();
+        prizeMoney.put(LottoNumberMatchCount.THREE, 5000);
```
로또 번호의 매치 카운트와 금액을 포함해서 enum 에서 관리해보시면 좋을 것 같습니다. :)
로또 게임을 해보시면 1등인 경우 6개의 번호가 일치하고, 20억의 상금(상금은 유동적이지만 미션의 제한 사항으로 20억으로 설정)을 가지고 있는 것을 알 수 있습니다.

따라서, enum 에서 등수를 관리한다면 아래와 같은 코드로 관리해보실 수 있을 것 같아요.
이렇게 하면 로또의 등수(도메인) 규칙을 한 클래스에서 관리하고, 이와 관련된 값을 하나의 클래스에서 관리할 수 있습니다.
각 등수에 대한 규칙(매치되어야 하는 개수, 상금)을 enum 내부에서 관리하므로 응집도가 올라간다고 볼 수 있습니다.
```java
public enum LottoRank {
    FIRST(6, 2_000_000_000)
    ;

    private final int matchCount;
    private final int prizeMoney;
}
```

다음 단계 미션을 진행하실 때 여유가 되신다면 위 코드를 참고하시어 `LottoNumberMatchCount` 부분을 리팩토링해보셔도 좋겠습니다.

OptimistLabyrinth replied:
수정 완료했습니다.

seokhongkim reviewed (approved):
기성님, 3단계 미션 코멘트 잘 반영해주셨습니다. 💯 
3단계 PR은 이만 머지하도록 하겠습니다.

코멘트 조금 남겼는데, 4단계 미션 진행하실 때 여유가 되시면 함께 고민해보셔도 좋겠습니다.
4단계 미션에서 뵙겠습니다. 😀
